### PR TITLE
[IMP] website_event: move event question and add identification questions

### PR DESCRIPTION
**Pull request by pko-odoo**
This pr moves the code from website_event_questions to website_event and from website_event_crm_questions to website_event_crm module.

And it changes the way the identification questions (name, email, phone) are asked when registering to an event.
They aren't hardcoded anymore and can be created per event the same way other questions can be.
They can be set as mandatory or not and the order can be changed.
One can now also ask for the attendee company name.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_event/models/event_event.py [hunk 1]
```diff
@@ -190,6 +211,51 @@ def _compute_website_url(self):
...
+        if self._origin.question_ids:
+            # lines to keep: those with already given answers
+            questions_tokeep_ids = self.env['event.registration.answer'].search(
+                [('question_id', 'in', self._origin.question_ids.ids)]
+            ).question_id.ids
+        else:
+            questions_tokeep_ids = []
+        for event in self:
+            if not event.event_type_id and not event.question_ids:
+                event.question_ids = self._default_question_ids()
```
Comment (tde-banana-odoo) [inline addons/website_event/models/event_event.py:234]: Note that this is not really in the original code being moved ... changing the default behavior should be done in its own commit, to ease traceability.

File: addons/website_event/models/event_question.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    _name = 'event.question'
+    _rec_name = 'title'
+    _order = 'sequence,id'
+    _description = 'Event Question'
+
+    title = fields.Char(required=True, translate=True, compute='_compute_title', readonly=False, store=True)
+    question_type = fields.Selection([
+        ('simple_choice', 'Selection'),
+        ('text_box', 'Text Input'),
+        ('name', 'Name'),
```
Comment (tde-banana-odoo) [inline addons/website_event/models/event_question.py:18]: Seems those types are new ... should be done in its own commit with explanations, in order to see how the model evolves.

File: addons/website_event/models/event_event.py [hunk 3]
```diff
@@ -88,6 +92,23 @@ def _default_cover_properties(self):
...
+        'event.question', 'event_id', 'General Questions',
+        domain=[('once_per_order', '=', True)])
+    specific_question_ids = fields.One2many(
+        'event.question', 'event_id', 'Specific Questions',
+        domain=[('once_per_order', '=', False)])
+
+    @api.constrains('question_ids')
+    def _check_unique_identification_question_ids(self):
+        for event in self:
+            question_types = event.question_ids.filtered(lambda q: q.is_identification_question).mapped('question_type')
```
Comment (tde-banana-odoo) [inline addons/website_event/models/event_event.py:109]: Isn't it possible to have constraints based on event_id / question_type directly at sql level on question model ?

File: addons/website_event/models/event_question.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        if any(question.event_type_id and question.event_id for question in self):
+            raise UserError(_('Question cannot belong to both the event category and itself.'))
+
+    @api.depends('question_type')
+    def _compute_title(self):
+        for question in self:
+            question_type_dict = dict(self._fields['question_type'].selection)
+            if question.question_type not in ['simple_choice', 'text_box']:
+                question.title = question_type_dict.get(question.question_type)
+            elif question.title in question_type_dict.values():
```
Comment (tde-banana-odoo) [inline addons/website_event/models/event_question.py:44]: Not sure to understand the purpose here ...  are we resetting the title if it is "Name" ? What is the added value ?

File: addons/website_event/models/event_question.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            question.is_identification_question = question.question_type in ['name', 'email', 'phone', 'company']
+
+    def write(self, vals):
+        """ We add a check to prevent changing the question_type of a question that already has answers.
+        Indeed, it would mess up the event.registration.answer (answer type not matching the question type). """
+
+        if 'question_type' in vals:
+            questions_new_type = self.filtered(lambda question: question.question_type != vals['question_type'])
+            if questions_new_type:
+                answer_count = self.env['event.registration.answer'].search_count([('question_id', 'in', questions_new_type.ids)])
```
Comment (tde-banana-odoo) [inline addons/website_event/models/event_question.py]: I know this is the original code, but can't we use 'answer_ids' field instead of doing a search_count ?

**Review by tde-banana-odoo — COMMENTED** [review]
First validation ! Looked at global models, added some comments :) as explained on task, first step will anyway be to split the mov from the ref/imp/code cleaning, to keep history and allow to better understand changes.

Once updated and current comments answered/done, will have another look :)

Thanks and cheers !

**Review by tde-banana-odoo — COMMENTED** [review]
Second check, but some questions arise during the validation, going back to the task to ask them :) .

File: addons/website_event/models/event_question.py [hunk 6]
```diff
@@ -23,12 +28,41 @@ class EventQuestion(models.Model):
...
+                question.title = False
+
+    @api.depends('question_type')
+    def _compute_is_identification_question(self):
+        for question in self:
+            question.is_identification_question = question.question_type in ['name', 'email', 'phone', 'company_name']
+
+    def init(self):
+        # Allows only 1 same identification question per event / event type
+        self.env.cr.execute("""
```
Comment (tde-banana-odoo) [inline addons/website_event/models/event_question.py:57]: Shoudln't we also have a constraint about "either event_id, either event_type_id" ?

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, I think it is ok for the validation, especially if we want it for the freeze ^^ few comments, but not blocking for testing :)

File: addons/website_event/tests/test_website_event.py [hunk 7]
```diff
@@ -29,6 +31,87 @@ def test_website_event_pages_seo(self):
         self.assertEqual(intro_event_menu.view_id.website_meta_title, "Hello, world!")
         self.assertEqual(event.website_meta_title, False)
 
+    def test_01_tickets_questions(self):
```
Comment (awa-odoo) [inline addons/website_event/tests/test_website_event.py:34]: Should we rename to "test_website_event_questions" to match naming of other tests?

**Review by awa-odoo — COMMENTED** [review]
Hello @pko-odoo 

Not fully tested yet but here is a preliminary review so we can try to squeeze this for the freeze.

I'm already submitting those comments as they are about the MOV part.
Will make a second batch with the comments on the other commit so it's easier for you.

Thanks & cheers!

File: addons/website_event/models/event_question.py (hunk 6 above)
Comment (pko-odoo) [reply addons/website_event/models/event_question.py:57]: Don't you mean like this https://github.com/odoo/odoo/blob/bf772181933ce5334da35c8368455963b2478399/addons/website_event_questions/models/event_question.py#L27 ? ^^

File: addons/website_event/views/event_templates_page_registration.xml [hunk 8]
```diff
@@ -281,65 +281,52 @@
...
-                                </div>
-                                <t t-if="event.specific_question_ids">
-                                    <div t-foreach="event.specific_question_ids" t-as="question" class="row" t-att-name="question.title">
-                                        <div class="col-lg-6 mt-2">
-                                            <t t-call="website_event.registration_event_question">
-                                                <t t-set="registration_index" t-value="counter"/>
+                    <t t-if="availability_check">
+                        <t t-set="counter" t-value="0"/>
+                        <t t-set="input_type" t-value="{'name': 'text', 'email': 'email', 'phone': 'tel', 'company_name': 'text'}"/>
+                        <t t-set="type_labels" t-value="dict(env['event.question']._fields['question_type'].selection)"/>
```
Comment (awa-odoo) [inline addons/website_event/views/event_templates_page_registration.xml:289]: Pretty sure those are "untranslated".

Wouldn't it be easier to keep the title field for "identification questions" in the form view and just initialize it if empty to what we want?

It would simplify and also allow end users to just change the questions label if they want to (so "Mobile" instead of "Phone" for example).

Probably to confirm with Luc.

File: addons/website_event/models/event_question.py [hunk 9]
```diff
@@ -23,12 +28,41 @@ class EventQuestion(models.Model):
...
+            CREATE UNIQUE INDEX IF NOT EXISTS event_unique_identification_question
```
Comment (awa-odoo) [inline addons/website_event/models/event_question.py:58]: Have to say that creates very ugly error messages on the form view :D

Can't we go for a custom api.constrains to control a bit the error messages?

File: addons/website_event/models/event_question.py [hunk 10]
```diff
@@ -23,12 +28,41 @@ class EventQuestion(models.Model):
...
+    def _compute_title(self):
+        """ Since title is mandatory and to prevent the user from entering it
+        for identification questions, it will be computed automatically. But we
+        also want it to reset automatically if the question_type changes."""
+        for question in self:
+            question_type_dict = dict(self._fields['question_type']._description_selection(self.env))
+            if question.question_type not in ['simple_choice', 'text_box']:
+                question.title = question_type_dict.get(question.question_type)
+            elif question.title in question_type_dict.values():
+                question.title = False
```
Comment (awa-odoo) [inline addons/website_event/models/event_question.py:48]: Have to say I also don't understand the purpose, this seems rather complicated just for a title.

Would say let people enter what they want and initialize maybe for the "identification questions"  if we want to. See other comment.

(Which makes me wonder if we even need a "is_identification_question" field).

File: addons/website_event/models/event_type.py [hunk 11]
```diff
@@ -8,6 +8,13 @@ class EventType(models.Model):
     _name = 'event.type'
     _inherit = ['event.type']
 
+    def _default_question_ids(self):
+        return [
+            (0, 0, {'title': 'Name', 'question_type': 'name', 'is_mandatory_answer': True}),
```
Comment (awa-odoo) [inline addons/website_event/models/event_type.py:13]: Note that those are not translated.

File: addons/website_event/views/event_templates_page_registration.xml [hunk 12]
```diff
@@ -281,65 +281,52 @@
...
-                                    <input class="d-none" type="text" t-attf-name="#{counter}-event_ticket_id" t-attf-value="#{ticket['id']}"/>
-                                </div>
-                                <t t-if="event.specific_question_ids">
-                                    <div t-foreach="event.specific_question_ids" t-as="question" class="row" t-att-name="question.title">
-                                        <div class="col-lg-6 mt-2">
-                                            <t t-call="website_event.registration_event_question">
-                                                <t t-set="registration_index" t-value="counter"/>
+                    <t t-if="availability_check">
+                        <t t-set="counter" t-value="0"/>
+                        <t t-set="input_type" t-value="{'name': 'text', 'email': 'email', 'phone': 'tel', 'company_name': 'text'}"/>
```
Comment (awa-odoo) [inline addons/website_event/views/event_templates_page_registration.xml:288]: Would name "input_type_by_question_type", a bit more explicit.

File: addons/website_event/views/event_templates_page_registration.xml [hunk 13]
```diff
@@ -281,65 +281,52 @@
...
+                        <t t-set="type_labels" t-value="dict(env['event.question']._fields['question_type'].selection)"/>
+                        <t t-foreach="tickets" t-as="ticket">
+                            <t t-foreach="range(1, ticket['quantity'] + 1)" t-as="att_counter" name="attendee_loop">
+                                <t t-set="counter" t-value="counter + 1"/>
+                                <div class="modal-body">
+                                    <h5 t-attf-class="mt-1 pb-2 #{'border-bottom' if event.question_ids else ''}">Ticket #<span t-out="counter"/> <small class="text-muted">- <span t-out="ticket['name']"/></small></h5>
+                                    <t t-if="event.specific_question_ids">
+                                        <div class="row">
```
Comment (awa-odoo) [inline addons/website_event/views/event_templates_page_registration.xml:296]: Note that you can put the t-if into the div to same some spacing and indentation levels.

File: addons/website_event/tests/test_website_event.py [hunk 14]
```diff
@@ -48,6 +48,15 @@ def test_website_event_questions(self):
...
             'website_published': True,
             'question_ids': [(0, 0, {
+                'title': 'Name',
+                'question_type': 'name',
+            }), (0, 0, {
+                'title': 'Email',
+                'question_type': 'email',
+            }), (0, 0, {
+                'title': 'Phone',
+                'question_type': 'phone',
```
Comment (awa-odoo) [inline addons/website_event/tests/test_website_event.py:58]: Would be nice to add "company_name" in there and adapt the tour accordingly.

File: addons/website_event/tests/test_event_internals.py [hunk 15]
```diff
@@ -25,10 +25,12 @@ def test_event_type_configuration_from_type(self):
 
         self.assertEqual(
             event.question_ids.mapped('question_type'),
-            ['simple_choice', 'simple_choice', 'text_box'])
-        self.assertEqual(event.specific_question_ids.title, 'Question1')
+            ['name', 'email', 'phone', 'simple_choice', 'simple_choice', 'text_box'])
+        self.assertEqual(event.specific_question_ids.filtered(
```
Comment (awa-odoo) [inline addons/website_event/tests/test_event_internals.py:29]: You could also adapt to test the title of the added questions (name / phone / email) ?

**Review by awa-odoo — COMMENTED** [review]
Hello @pko-odoo 

Finished a first reading of the second commit.

Not a fan of the title magic with the "identification questions", would simplify that part.
See review comment for details and suggestions.

Rest is pretty straightforward, added some general comments about naming / test coverage / ...

Thanks for your work & cheers!

File: addons/website_event/views/event_templates_page_registration.xml [hunk 16]
```diff
@@ -350,6 +334,10 @@
 <template id="registration_event_question" name="Registration Event Question">
     <label t-out="question.title"/>
     <span t-if="question.is_mandatory_answer">*</span>
+    <t t-if="question.is_identification_question">
```
Comment (awa-odoo) [inline addons/website_event/views/event_templates_page_registration.xml:337]: Looks like "is_identification_question" is only used in tests and here in the end.

Can't we instead use input_type_by_question_type.get(question.question_type, 'some_default_value') ?

And in tests adapt a bit so that we don't need this field? Feels a bit overkill.

File: addons/website_event/controllers/main.py [hunk 17]
```diff
@@ -270,51 +264,59 @@ def _process_attendees_form(self, event, form_details):
             if int(ticket_id) not in event.event_ticket_ids.ids and len(event.event_ticket_ids.ids) > 0:
                 raise UserError(_("This ticket is not available for sale for this event"))
         registrations = {}
-        global_values = {}
         general_answer_ids = []
+        general_identification_answers = {}
+        already_treated_fields_data = {}
```
Comment (awa-odoo) [inline addons/website_event/controllers/main.py:269]: Would add a small comment for this.

Something along the lines of:

```
# as we may have several questions populating the same field (e.g: the phone)
# we use this to hold the fields that have already been handled
# goal is to use the answer to the first question of every 'type' (aka name / phone / email / company name)
```

File: addons/website_event/controllers/main.py [hunk 18]
```diff
@@ -270,51 +264,59 @@ def _process_attendees_form(self, event, form_details):
...
-                    registrations.setdefault(registration_index, dict())\
-                        .setdefault('registration_answer_ids', list()).append((0, 0, answer_values))
-            else:
-                counter, attr_name = key.split('-', 1)
-                field_name = attr_name.split('-')[0]
+            if not value:
+                continue
+
+            key_values = key.split('-')
+            if len(key_values) == 2:
```
Comment (awa-odoo) [inline addons/website_event/controllers/main.py:275]: Same idea, would add a small comment mentioning that we have a special use case for the ticket id.

File: addons/website_event/controllers/main.py [hunk 19]
```diff
@@ -270,51 +264,59 @@ def _process_attendees_form(self, event, form_details):
...
+                continue
+
+            registration_index, question_type, question_id = key_values
+            answer_values = None
+            if question_type == 'simple_choice':
+                answer_values = {
+                    'question_id': int(question_id),
+                    'value_answer_id': int(value)
+                }
+            elif question_type != 'event_ticket_id':
```
Comment (awa-odoo) [inline addons/website_event/controllers/main.py:287]: If I'm not mistaken we don't need this anymore, right? Could be a simple "else"?

File: addons/website_event/controllers/main.py (hunk 18 above)
Comment (awa-odoo) [inline addons/website_event/controllers/main.py:275]: Hmm actually I think the previous code was checking if the field_name was inside "_get_website_registration_allowed_fields", and here we don't do it anymore.

We could probably just re-add the check here as we now have 2 separate cases:
- "special" fields that are not questions, aka partner_id, event_id, event_ticket_id
- fields added through questions of type 'name', 'email', 'phone, 'company_name'

File: addons/website_event/controllers/main.py [hunk 20]
```diff
@@ -270,51 +264,59 @@ def _process_attendees_form(self, event, form_details):
...
+                    'value_text_box': value
+                }
+
+            if answer_values and not int(registration_index):
+                general_answer_ids.append((0, 0, answer_values))
+            elif answer_values:
+                registrations.setdefault(registration_index, dict())\
+                    .setdefault('registration_answer_ids', list()).append((0, 0, answer_values))
+
+            field_name = question_type
```
Comment (awa-odoo) [inline addons/website_event/controllers/main.py:299]: Would assign field_name after checking that the question_type is in our list.

File: addons/website_event/tests/test_event_internals.py (hunk 15 above)
Comment (awa-odoo) [reply addons/website_event/tests/test_event_internals.py:29]: Up?

File: addons/website_event/views/event_templates_page_registration.xml [hunk 21]
```diff
@@ -281,65 +281,49 @@
...
-                                            t-att-value="default_first_attendee.get('email', '') if counter == 1 else ''"/>
-                                    </div>
-                                    <div class="col-lg my-2">
-                                        <label>Phone <small>(Optional)</small></label>
-                                        <input class="form-control" type="tel" t-attf-name="#{counter}-phone"
-                                            t-att-value="default_first_attendee.get('phone', '') if counter == 1 else ''"/>
+                    <t t-if="availability_check">
+                        <t t-set="counter" t-value="0"/>
+                        <t t-set="input_type_by_question_type" t-value="{'name': 'text', 'email': 'email', 'phone': 'tel', 'company_name': 'text'}"/>
+                        <t t-foreach="tickets" t-as="ticket">
```
Comment (awa-odoo) [inline addons/website_event/views/event_templates_page_registration.xml:289]: Could probably keep the t-if inside the t-foreach as I believe this increases the diff by a lot? As we have an extra indentation level everywhere?

File: addons/website_event/tests/test_event_internals.py [hunk 22]
```diff
@@ -56,18 +61,18 @@ def test_process_attendees_form(self):
...
-            '1-event_ticket_id': ticket_id_1.id,
-            '2-name': 'Geluchat',
-            '2-email': 'geluchat@gmail.com',
-            '2-phone': '+32777777777',
-            '2-event_ticket_id': ticket_id_2.id,
-            'question_answer-1-%s' % self.event_question_1.id: '5',
-            'question_answer-2-%s' % self.event_question_1.id: '9',
-            'question_answer-0-%s' % self.event_question_2.id: '7',
-            'question_answer-0-%s' % self.event_question_3.id: 'Free Text',
+            '1-name-%s' % self.event_identification_question_1.id: 'Pixis',
```
Comment (awa-odoo) [inline addons/website_event/tests/test_event_internals.py:64]: As we have a heuristic to test that we pick the first value for multiple identification questions, could be interesting to have 2 phone questions for example, and make sure we populate the field with the first answer.

Would also be nice to include the company_name into this.

**Review by awa-odoo — COMMENTED** [review]
Hello @pko-odoo 

Last bits of polishing and then we should be good to go :rocket: 

Thanks for keeping up with me on this & cheers!

**Comment by awa-odoo** [comment]
( @pko-odoo Looks like runbot is still exploding by the way so make sure to re-run all tests locally and fix them :slightly_smiling_face: )

File: addons/website_event/tests/test_event_internals.py (hunk 15 above)
Comment (pko-odoo) [reply addons/website_event/tests/test_event_internals.py:29]: The test is a few lines below ^^

**Comment by awa-odoo** [comment]
@robodoo r+ rebase-ff

**Comment by awa-odoo** [comment]
@robodoo r+

**Comment by awa-odoo** [comment]
@robodoo r+

**Comment by awa-odoo** [comment]
@robodoo r+
